tensorflow/compiler/mlir/tensorflow/ir/tf_ops_n_z.cc

```diff
@@ -947,9 +947,42 @@ LogicalResult ShapeNOp::fold(ArrayRef<Attribute> operands,
   return success();
 }
 
-// TODO(hinsu): Add canonicalization pattern for ShapeN ops that don't have all
+namespace {
+// Canonicalization pattern for ShapeNOp that don't have all
 // static input shapes. Replacing output values corresponding to static input
 // types may enable optimizations in users of the values.
+class ShapeNPartialStaticInputShape : public OpRewritePattern<ShapeNOp> {
+  using OpRewritePattern<ShapeNOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(ShapeNOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.getNumOperands() == 0) return success();
+    int width = op.getType(0)
+                    .cast<ShapedType>()
+                    .getElementType()
+                    .getIntOrFloatBitWidth();
+    BoolAttr use32Bit = BoolAttr::get(width == 32, op.getContext());
+
+    SmallVector<Value, 4> results;
+    for (Value input : op.getOperands()) {
+      Value shape;
+      if (OpFoldResult result = ConvertShapeToAttr(input.getType(), width)) {
+        shape =
+            rewriter.create<TF::ConstOp>(op.getLoc(), result.get<Attribute>());
+      } else {
+        shape = rewriter.create<TF::ShapeOp>(op.getLoc(), input, use32Bit);
```

Comment:
Instead of creating individual Shape ops, we should create a single ShapeN op with all the dynamic inputs.

Reply:
Updated! A quick followup here. Should we canonicalize `ShapeNOp` to `ShapeOp` when n = 1?

Reply:
Yes, that's a good idea.

Reply:
> Updated! A quick followup here. Should we canonicalize `ShapeNOp` to `ShapeOp` when n = 1?